kvs-watch: Support lookup with WAITCREATE flag and w/o need for WATCH flag

chu11 opened the pull request:
Per discussion in #1828, this PR allows users to specify the WAITCREATE flag without specifying the WATCH flag.  This allows users to do a basic lookup that waits for a key to be created before returning.

Given the minimal changes necessary to support this, I was initially hesitant to believe that I had done this correctly.  But I think this is all that's necessary.

garlick commented on src/common/libkvs/kvs_lookup.c:
```diff
@@ -126,7 +134,8 @@ flux_future_t *flux_kvs_lookup (flux_t *h, int flags, const char *key)
     if (!(ctx = alloc_ctx (h, flags, key)))
         return NULL;
-    if (flags & FLUX_KVS_WATCH)
+    if (flags & FLUX_KVS_WATCH
```
Parens around the bitwise & operations are usually advised here.  (hmm, thought `-Werror` would flag this as potentially confusing).

garlick commented on src/modules/kvs-watch/kvs-watch.c:
```diff
@@ -470,6 +470,14 @@ static void lookup_continuation (flux_future_t *f, void *arg)
+         * flag to indicate that break out condition.
+         */
+        if (w->responded
+            && w->flags & FLUX_KVS_WAITCREATE
```
Parens advised, as above.

garlick commented on doc/man1/flux-kvs.adoc:
```diff
@@ -68,22 +68,24 @@ Remove a kvs namespace.
-monitored, which may miss several unique situations, such as the
-replacement of an entire parent directory.  The '-f' option can be
-specified to monitor for many of these special situations.
+is specified, until 'count' values have been displayed.  The '-W' can
```
This sentence
> The -W can be used in conjunction with -w to wait for a key to be created.

maybe should be removed?

chu11 replied:
I had debated that sentence.  I'll just axe it.

garlick commented on t/t1007-kvs-lookup-watch.t:
```diff
@@ -143,6 +143,28 @@ test_expect_success NO_CHAIN_LINT 'flux kvs get --watch and --uniq do not see du
+        pid=$! &&
+        wait_watcherscount_nonzero primary &&
+        flux kvs put test.waitcreate2=2 &&
+	$waitfile --count=1 --timeout=10 \
```
Shouldn't thiss just be a `flux kvs get --waitcreate test.waitcreate`?

chu11 replied:
Ahhh, I see, use that instead of waitfile, but still do the check on the prior get.  Good idea.

chu11 replied:
oh wait, I need the waitfile call to wait on the background process to complete, otherwise the test can be racy.  Or were you thinking of something else?

chu11 replied:
ahh good catch, I'll add that.  It's sort of covered with the --watch & --waitcreate tests, but it's good to cover that too.

garlick commented:
This looks good.  I tried running `flux kvs get --waitcreate foo`, then in another window, `flux kvs put foo=bar`, and that worked as expected.

Next I tried the same, but on a non-default namespace that didn't exist when the `get` was started, and it worked as expected.

So victory I think!

chu11 commented:
re-pushed and went ahead and squashed tiny fixes.  Added some fixes for parens around bitwise ops that were missed in prior patches.

I didn't fix the test in discussion above, b/c I think it should be kept way it is.  Unless I'm missing something about the suggestion.

chu11 commented:
repushed with the extra --waitcreate test suggested

chu11 commented:
re-pushed, forgot some NO_CHAIN_LINTs (not sure how it passed before though).

garlick commented:
Nice!  Thanks @chu11!
